onnx/onnx-ml.proto

```diff
@@ -384,10 +384,39 @@ message TensorShapeProto {
       int64 dim_value = 1;
       string dim_param = 2;   // namespace Shape
     };
+    // Standard denotation can optionally be used to denote tensor
+    // dimensions with standard semantic descriptions to ensure
+    // that operations are applied to the correct axis of a tensor.
+    optional string standard_denotation = 3;
   };
   repeated Dimension dim = 1;
 }
 
+// A set of pre-defined constants to be used as values for
+// the standard denotation field in TensorShapeProto.Dimension
+// for semantic description of the tensor dimension.
+message StandardDenotationConstProto {
+  // Describe a batch number dimension.
+  optional string DATA_BATCH = 1 [default = "DATA_BATCH"];
+  // Describe a channel dimension.
+  optional string DATA_CHANNEL = 2 [default = "DATA_CHANNEL"];
+  // Describe a time dimension.
+  optional string DATA_TIME = 3 [default = "DATA_TIME"];
+  // Describe a feature dimension. This is typically a feature
+  // dimension in RNN and/or spatial dimension in CNN.
+  optional string DATA_FEATURE = 4 [default = "DATA_FEATURE"];
+  // Describe a filter in-channel dimension. This is the dimension
+  // that is identical to the channel dimension of the input
+  // image feature maps.
+  optional string FILTER_IN_CHANNEL = 5 [default = "FILTER_IN_CHANNEL"];
```

Comment:
actually if denotations go through shape inference  - shouldn't this be just FILTER_CHANNEL? otherwise it's weird that IN_ becomes OUT_ when passed to the next op

or is it only for the weights parameter?

Reply:
Yes, it's only for the weights parameter thus I do not expect this to be propagated. I sort of see the point of confusion. Now we are adding semantic description to the definition of a dimension thus the comment should be "identical in size" since they are not identical in meaning.